### PR TITLE
adding creation_time and last_update_time to experiment

### LINKED ISSUE #6462: [FR] Implement `last_update_time` and `creation_time` for experiments
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

Documentation for mlflow experiments returned via the REST API refers to `last_update_time` and `creation_time` fields https://www.mlflow.org/docs/latest/rest-api.html#mlflowexperiment, but these values are never returned in the response.

Under the hood, these fields appear to be missing from the [experiment entity](https://github.com/mlflow/mlflow/blob/master/mlflow/entities/experiment.py) and are only defined in the [protobuf message](https://github.com/mlflow/mlflow/blob/174c81f14364425f00fb9deca789fa5435446c0f/mlflow/protos/service.proto#L668-L672).

Would it be possible to implement these fields as the documentation suggests?

### Motivation

> #### What is the use case for this feature?
* Parity between docs and implementation - the docs already imply the existence of this functionality

> #### Why is this use case valuable to support for MLflow users in general?
* They won't be misled by the docs

> #### Why is this use case valuable to support for your project(s) or organization?
* We are attempting to export mlflow experiment/run data via KafkaConnect into a separate system with more powerful visualization functionality. The lack of `last_update_time` means that you can't easily detect which experiments have been updated since the last run.

> #### Why is it currently difficult to achieve this use case?
* No timestamp fields on experiment entities

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by subramaniam02**
## Related Issues/PRs

closes #6462 

## What changes are proposed in this pull request?

`creation_time` and `last_update_time` attributes  added to experiment

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by subramaniam02** [comment]
@harupy Sorry about creating a new PR. Please take a look.

File: tests/store/tracking/test_file_store.py [hunk 1]
```diff
@@ -402,10 +408,14 @@ def test_create_experiment(self):
             exp1.artifact_location,
             path_to_local_file_uri(posixpath.join(self.test_root, created_id)),
         )
+        assert isinstance(exp1.creation_time, int)
+        assert isinstance(exp1.last_update_time, int)
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:412]:
```suggestion
        assert exp1.creation_time >= time_before_create 
        assert exp1.creation_time == exp1.last_update_time
```

Can we make this more strict?

File: mlflow/store/tracking/file_store.py [hunk 2]
```diff
@@ -433,6 +440,14 @@ def delete_experiment(self, experiment_id):
                 "Could not find experiment with ID %s" % experiment_id,
                 databricks_pb2.RESOURCE_DOES_NOT_EXIST,
             )
+        experiment = self._get_experiment(experiment_id)
+        experiment._set_last_update_time(int(time.time() * 1000))
+        meta_dir = os.path.join(self.root_directory, experiment_id)
+        FileStore._overwrite_yaml(
+            root=meta_dir,
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:447]:
```suggestion
        experiment_dir = os.path.join(self.root_directory, experiment_id)
        FileStore._overwrite_yaml(
            root=experiment_dir,
```

nit

File: tests/store/tracking/test_file_store.py [hunk 3]
```diff
@@ -402,10 +408,14 @@ def test_create_experiment(self):
...
 
         # get the new experiment (by name) and verify (by id)
         exp2 = fs.get_experiment_by_name(name)
         self.assertEqual(exp2.experiment_id, created_id)
+        assert isinstance(exp2.creation_time, int)
+        assert isinstance(exp2.last_update_time, int)
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:418]: https://github.com/mlflow/mlflow/pull/6756/files#r968032645 same here

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -528,6 +531,8 @@ def test_create_experiments(self):
         actual = self.store.get_experiment(experiment_id)
         self.assertEqual(actual.experiment_id, experiment_id)
         self.assertEqual(actual.name, "test exp")
+        assert isinstance(actual.creation_time, int)
+        assert isinstance(actual.last_update_time, int)
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:535]: same here https://github.com/mlflow/mlflow/pull/6756/files#r968032645

**Review by harupy — APPROVED** [review]
@subramaniam02 Left comments for improving tests, otherwise LGTM!

File: tests/store/tracking/test_file_store.py (hunk 1 above)
Comment (subramaniam02) [reply tests/store/tracking/test_file_store.py]: done

File: mlflow/store/tracking/file_store.py [hunk 5]
```diff
@@ -337,7 +337,14 @@ def _create_experiment_with_id(self, name, experiment_id, artifact_uri, tags):
...
         meta_dir = mkdir(self.root_directory, str(experiment_id))
-        experiment = Experiment(experiment_id, name, artifact_uri, LifecycleStage.ACTIVE)
+        experiment = Experiment(
+            experiment_id,
+            name,
+            artifact_uri,
+            LifecycleStage.ACTIVE,
+            creation_time=int(time.time() * 1000),
+            last_update_time=int(time.time() * 1000),
+        )
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:347]:
```suggestion
        creation_time = time.time() * 1000
        experiment = Experiment(
            experiment_id,
            name,
            artifact_uri,
            LifecycleStage.ACTIVE,
            creation_time=creation_time,
            last_update_time=creation_time,
        )
```

File: tests/store/tracking/test_file_store.py [hunk 6]
```diff
@@ -387,7 +393,8 @@ def test_create_experiment(self):
             fs.create_experiment(None)
         with pytest.raises(Exception, match="Invalid experiment name: ''"):
             fs.create_experiment("")
-
+        test_case_start_time = time.time() * 1000
+        time.sleep(0.5)
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:397]: I think we can remove this `sleep`.

File: tests/store/tracking/test_file_store.py [hunk 7]
```diff
@@ -402,10 +409,14 @@ def test_create_experiment(self):
             exp1.artifact_location,
             path_to_local_file_uri(posixpath.join(self.test_root, created_id)),
         )
+        assert exp1.creation_time > test_case_start_time
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:412]:
```suggestion
        assert exp1.creation_time >= test_case_start_time
```

Comment (harupy) [reply tests/store/tracking/test_file_store.py:412]: In delete/restore/rename operation, we need sleep to ensure `last_update_time` has been updated, but in create operation, we don't.

File: mlflow/store/tracking/file_store.py (hunk 5 above)
Comment (harupy) [reply mlflow/store/tracking/file_store.py:347]: Can we apply the same change to the SQLAlchemy store?

File: tests/store/tracking/test_file_store.py [hunk 8]
```diff
@@ -387,7 +393,8 @@ def test_create_experiment(self):
             fs.create_experiment(None)
         with pytest.raises(Exception, match="Invalid experiment name: ''"):
             fs.create_experiment("")
-
+        test_case_start_time = time.time() * 1000
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:396]: Can we move this line to line 388 or rename the variable?

File: tests/store/tracking/test_file_store.py [hunk 9]
```diff
@@ -476,18 +487,28 @@ def test_delete_restore_experiment(self):
         exp_id = self.experiments[random_int(0, len(self.experiments) - 1)]
         exp_name = self.exp_data[exp_id]["name"]
 
+        exp1 = fs.get_experiment(exp_id)
+        time.sleep(0.05)
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:491]:
```suggestion
        time.sleep(0.01)
```

Let's use `0.01` to run tests faster.

**Comment by harupy** [comment]
@subramaniam02 I left a few commits to resolve my comments. Thanks for the contribution!

**Pull request merged**